### DISCUSSION TITLE
Параллельная работа двух аккаунтов Spotify

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#251] [Помощь с алгоритмом]
В целях решения вопроса с новыми лайками-галочками, решил опробовать идею использования двух аккаунтов Spotify. Проблема этих галочек в том, что они показывают, что трек добавлен в медиатеку, но не показывают куда именно, в лайки и/или в какие-то другие плейлисты. Поэтому в своих плейлистах, в том числе и созданных с помощью Goofy, совершенно не понятно лайкнут трек, или нет.
Суть идеи такова, что все плейлисты создаются на аккаунте-дублере Spotify, а прослушиваются на основном аккаунте. Это позволяет иметь на основном аккаунте всего один свой плейлист с лайками, и соответственно видеть в остальных сгенерированных плейлистах какие треки лайкнуты.

По [инструкции](https://chimildic.github.io/goofy/#/best-practices?id=%d0%9d%d0%b5%d1%81%d0%ba%d0%be%d0%bb%d1%8c%d0%ba%d0%be-%d0%b0%d0%ba%d0%ba%d0%b0%d1%83%d0%bd%d1%82%d0%be%d0%b2) создал копию Goofy и подключил к аккаунту дублеру. Вроде все ок, работает как и положено. 
Но теперь встал вопрос, как это добро использовать параллельно. Просто запускать те же скрипты для создания плейлистов на аккаунте дублере не получится, т.к. на нем невозможно использовать такие важные данные основного аккаунта как история, лайки и т.п. 
Логичней всего было бы продолжать создавать плейлисты на основном аккаунте, но получать результат  не в виде плейлистов, а в виде файлов кэша. А уже на аккаунте дублере из этих файлов создавать плейлисты. Собственно и вопрос - это вообще возможно и как это реализовать, если да? Если нет, то какие еще возможны способы реализации идеи?
Попутно вопрос, возможна ли вообще какая-то взаимосвязанная работа этих двух аккаунтов? Например запуск функции одного аккаунта, командой из другого.

**Comment by Chimildic** [comment]
upd: данный способ не рекомендуется. Лучше используйте [инструкцию про несколько аккаунтов](https://chimildic.github.io/goofy/#/best-practices?id=Несколько-аккаунтов).

## Старый пост 
>  использовать такие важные данные основного аккаунта как история, лайки

Кэш общий для всех проектов, если знать путь до файла. Последние примеры [отсюда](https://chimildic.github.io/goofy/#/best-practices?id=google-drive). Например, чтение истории другого аккаунта `Cache.read('root/id аккаунта/SpotifyRecentTracks')`

> взаимосвязанная работа этих двух аккаунтов

Можно переключать активный токен в одном проекте.

### Настройка

1. Надо узнать id аккаунтов. Они есть на странице профиля в адресе страницы. Например, для `https://open.spotify.com/user/abcd?query=example` id это `abcd`

![image](https://github.com/Chimildic/goofy/assets/40138097/3e76f75d-e797-4ab3-b997-ac8d873ee466)

2. Добавляем эти id в config файл в виде констант

```
const MAIN_SPOTIFY_ID = 'main_id'
const SECOND_SPOTIFY_ID = 'child_id'

function setProperties() {
    // ...
}
```

3. В основном файле библиотеки goofy находим модуль `Auth` (ближе к концу) и заменяем его целиком на новый

```
const Auth = (function () {
    const SCOPE = [
        'user-read-private', 'user-read-recently-played', 'user-read-currently-playing', 'user-read-playback-position',
        'user-read-playback-state', 'user-modify-playback-state', 'user-library-read', 'user-library-modify',
        'user-top-read', 'user-follow-read', 'user-follow-modify', 'playlist-read-private', 'playlist-modify-private',
        'playlist-modify-public', 'ugc-image-upload',
    ];
    const serviceMap = {}
    let serviceName = `spotify.${UserProperties.getProperty('userId')}`

    return {
        reset, hasAccess, getAccessToken, displayAuthPage, displayAuthResult, activeService
    };

    function activeService(userId) {
      serviceName = `spotify.${userId}`
      KeyValue['userId'] = userId
      UserProperties.setProperty('userId', userId)
    }

    function createService(name) {
        return OAuth2.createService(name)
            .setAuthorizationBaseUrl('https://accounts.spotify.com/authorize')
            .setTokenUrl('https://accounts.spotify.com/api/token')
            .setClientId(KeyValue.CLIENT_ID)
            .setClientSecret(KeyValue.CLIENT_SECRET)
            .setCallbackFunction('displayAuthResult_')
            .setPropertyStore(UserProperties)
            .setScope(SCOPE)
            .setParam('response_type', 'code')
            .setParam('redirect_uri', getRedirectUri());
    }

    function displayAuthResult(request) {
        let isAuthorized = getOrCreateService().handleCallback(request);
        return HtmlService.createHtmlOutput(isAuthorized ? 'Успешно!' : 'Отказано в доступе');
    }

    function displayAuthPage() {
        let template = '<a href="%s" target="_blank">Выдать права доступа</a><p>%s</p>';
        let html = Utilities.formatString(template, getOrCreateService().getAuthorizationUrl(), getRedirectUri());
        return HtmlService.createHtmlOutput(html);
    }

    function getRedirectUri() {
        let scriptId = encodeURIComponent(ScriptApp.getScriptId());
        return `https://script.google.com/macros/d/${scriptId}/usercallback`;
    }

    function hasAccess() {
        return getOrCreateService().hasAccess();
    }

    function getAccessToken() {
        return getOrCreateService().getAccessToken();
    }

    function reset() {
        getOrCreateService().reset();
    }

    function getOrCreateService() {
      if (serviceMap[serviceName] == undefined) {
        serviceMap[serviceName] = createService(serviceName)
      }
      return serviceMap[serviceName]
    }
})();
```

4. В самом верху того же файла есть функция `doGet`. Заменяем ее на следующую: 

```
function doGet() {
    Auth.activeService(SECOND_SPOTIFY_ID)
    return Auth.hasAccess() ? displayLaunchPage_() : Auth.displayAuthPage();
}
```

5. Идем в пробные развертывания и проходим авторизацию
6. Меняем в `doGet` константу `SECOND_SPOTIFY_ID` на `MAIN_SPOTIFY_ID` и снова проходим авторизацию (повторить для всех аккаунтов, если их больше двух, последний всегда `MAIN_SPOTIFY_ID`)

### Пример использования

После настройки конфиг содержит токен для каждого аккаунта. Функция `Auth.activeService` указывает токен какого аккаунта использовать. По умолчанию выбран тот, что последний раз использовался в `doGet` (она вызывается в том числе веб-интерфейсом)

```
function twoAccountCase() {
  // Работа в контексте главного аккаунта
  Auth.activeService(MAIN_SPOTIFY_ID)
  let savedTracksMain = Source.getSavedTracks(10)
  // ...

  // Переключение на второй аккаунт
  Auth.activeService(SECOND_SPOTIFY_ID)
  let savedTracksSecond = Source.getSavedTracks(10)
  // ...

  // Вернуть аккаунт по умолчанию
  Auth.activeService(MAIN_SPOTIFY_ID)
}
```

У такого способа большая проблема из-за постоянных переключений. Запуск очередной функции может наложится на исполнение другой функции, которая сменила активный токен. В целом значения кэшируются в локальных переменных на время исполнения. Поэтому важно только начало и конец исполнения. Везде первой сточкой активировать главный токен `Auth.activeService(MAIN_SPOTIFY_ID)`, а где токен менялся еще и возвращать главный обратно (пример выше `twoAccountCase`). 

```
// Любая функция, где только один токен
function any() {
  Auth.activeService(MAIN_SPOTIFY_ID)
  // ...
}
```

Другая проблема это кэш. Без явного указания пути, все пишется в папку с id аккаунта, который был по умолчанию на момент старта. Учитывая как часто запускается триггер истории, легко попасть в момент переключенного токена и намешать "чужих" данных или пропустить оригинальные. Но можно и модуль `Cache` пропатчить, чтоб динамически папку отдавал. 

Возможно что-то еще забыл, только тестировать остается. Первый вариант с общим кэшем мне нравится больше)

**Reply by Bedmit** [reply]
Сперва огромное спасибо за ответ. И не просто теоретический, а как всегда сразу с готовым решением!

Протестировал в течении дня вариант с простым чтением из файла кэша. Проще всего оказалось перевести все требуемые плейлисты на чтение в дублере из папок основного аккаунта. На 90% это покрыло все мои потребности. 

С переключениями токенов что-то побоялся связываться. 
По большому счету запуск скрипта аккаунта дублера из основного аккаунта требуется только на одном плейлисте. Но часто. Сам скрипт запускается Таскером на основном аккаунте, при переходе с одного трека на другой. После прохождения проверки того, что в плейбэке нужный плейлист, скрипт его обновляет и записывает в файл. Нужно было чтобы после этого скрипт аккаунта дублера считав этот файл обновил плейлист у себя. 
Сначала решил второй аккаунт тоже привязать к Таскеру и запускать его оттуда. Но сходу не разобрался как это сделать, чтобы не отвязать при этом первый. Да и тоже не хочется таких сложностей.

Пока скрипт в дублере просто запускается каждые пять минут по триггеру. Пока идет прослушивание плейлиста, все работает как надо. Хочу дополнительно создать условия при которых плейлист не будет перезаписываться вхолостую при простое. К тому же в такой скрипт можно добавить проверку и некоторых других плейлистов. Только не хватает элементарной грамотности. 

Суть идеи: в требуемых скриптах основного аккаунта, при их успешном обновлении в файл записывается некоторая константа.
Например в скрипте Smartify  `Cache.write('StartScript.json', 1);` 
в скрипте Bedtime `Cache.write('StartScript.json', 2);` 
в скрипте Flow `Cache.write('StartScript.json', 3);` 
В аккаунте дублере каждые пять минут запускается скрипт проверки этих констант. Из файла считывается константа и в зависимости от ее значения запускается соответствующая одноименная функция в дублере. В конце каждой функции константа обнуляется, что бы если при проверке был ноль, то скрипт проверки заканчивался ни чего не запуская. 
Вот как раз как этот проверочный скрипт написать ума не хватает. Возможно вместо констант лучше сразу использовать названия самих функций, чтобы не цифры проверять, а сразу запускать требуемую функцию.

Попутно вопрос. В дублере мне совершенно не нужны ни какие его данные. Ни история, ни лайки. Может логично триггер runTasks_ вообще отключить для экономии?

**Comment by Chimildic** [comment]
> Может логично триггер runTasks_ вообще отключить

Да, можно удалить триггер и закомментировать первые строчки в модуле `Clerk`, чтоб триггер не восстанавливался. 

```js
const Clerk = (function () {
    let tasks;
    // let functionName = 'runTasks_';
    // let taskTrigger = Trigger.get(functionName);
    // if (taskTrigger && taskTrigger.isDisabled()) {
    //     Trigger.remove(taskTrigger);
    //     taskTrigger = undefined;
    // }
    // if (!taskTrigger) {
    //     Trigger.create(functionName, 15);
    // }
    return {
        runOnceAfter, runOnceAWeek,
    }

// ...

}
```

> и в зависимости от ее значения запускается соответствующая одноименная функция в дублере

На самом деле можно запускать напрямую и смена токена не пригодится. 

1. Зайти в дублер и скопировать id проекта

![image](https://github.com/Chimildic/goofy/assets/40138097/71617286-f6a5-485a-aaf9-d0d2eadd331c)

2. Пойти в основной проект и добавить библиотеку по id

![image](https://github.com/Chimildic/goofy/assets/40138097/61a2cbef-92ca-4d37-a46a-df659630bd94)

3. Теперь по идентификатору библиотеки можно достучаться до публичных функций дублера

![image](https://github.com/Chimildic/goofy/assets/40138097/5322a7c2-2cb5-4048-aeaf-6c50e2b05f37)

### Пример использования

Чтобы создать плейлист в контексте дублера, нужна функция с принимающими аргументами. 

```js
// дублер
function saveWithReplace(params) {
    Playlist.saveWithReplace(params)
}
```

```js
// основа
function test() {
    let tracks = // ...
    // ...

    // Вместо DevGoofyVer183 имя библиотеки
    DevGoofyVer183.saveWithReplace({
        name: 'test',
        tracks: tracks
    })
}
```

**Reply by Bedmit** [reply]
Наконец то добрался до компа и  все протестировал. 
Работает все просто идеально! Это именно то что и было нужно. Без всяких токенов, кэша и таскера. Главное, что и сами скрипты не особо редактировать надо, просто в функции записи плейлиста заменить `Playlist`
 Вот только один нюанс. На основном аккаунте теперь не получается обновить управление развертыванием. Выдает ошибку
![2023-12-08_11-41-46](https://github.com/Chimildic/goofy/assets/76230194/d29e8c87-1efa-40d8-939a-743232a55e21)
Скрипты по триггеру обновляются, а вот с запуском через Таскер пока естественно ни как, т.к. не сохраняются изменения.

**Reply by Bedmit** [reply]
Добавился еще один нюансик. На основном аккаунте не сохраняется триггер. 
![2023-12-08_12-11-16](https://github.com/Chimildic/goofy/assets/76230194/b33f9018-214f-4aa7-a69d-885881d239ab)
Естественно ни каких блокировщиков не стоит. В аккаунте дублера все с добавлениями нормально.

**Reply by Bedmit** [reply]
Как я понял, все блокируется этим 
![2023-12-08_12-42-02](https://github.com/Chimildic/goofy/assets/76230194/aabb05ff-e5ad-4628-acf0-b23475cbd07c)
Если переключится с режима разработчика на 1, то и триггеры редактируются и развертывание обновляется. Но не работает запись плейлистов через дублер. Вернулся обратно, все заработало, кроме скриптов запускаемых Таскером. Выдает ошибку 
```
TypeError: CopyofGoofy.saveWithReplace is not a function
    at Smartify(Smartify:46:15)
```
Попробовал даже удалить библиотеку дублера и заново скопировав ее код ID добавить опять. Не помогло. Все работает кроме того что запускаю извне Таскером

**Reply by Chimildic** [reply]
После добавления в дублер функции `saveWithReplace` создать новое развертывание и в основе выбрать его (а не режим разработчика), а также обновить развертывание основы после этого. Так будет работать? 
Еще вариант сделать развертывание дублера публичным (при создании есть поле только для меня).

**Reply by Bedmit** [reply]
Все сделал. Первый вариант, без публичного развертывания. Все работает как часы! Как всегда огромное спасибо!
Протестирую немного. Может еще какие мысли появятся. Но пока просто идеально все.
